feat(docs): Historical trees docs (#3895)
AztecProtocol/aztec-packages: docs/docs/dev_docs/contracts/syntax/historical_access/how_to_access_history.md

---
title: How to access history from a smart contract
---

The Aztec Protocol uses an append-only Merkle tree to store the headers of all previous blocks in the chain as its leaves. This is known as an archive tree. You can learn more about how it works in the [concepts section](../../../../concepts/advanced/data_structures/trees.md#archive-tree).

# History library

> The history library allows you to access any of the following at a given block height before the current height:

Reviewer: ```suggestion
The history library allows you to prove any of the following at a given block height before the current height:
```

Technically you have to give the data to the library so it doesn't allow you to access it. It just allows you to prove the relevant statements if you provide it the relevant preimages.

s

> * [Private notes](#note-inclusion)

Reviewer: ```suggestion
* [Note inclusion](#note-inclusion)
```

> * [Notes that have been nullified](#nullifier-inclusion)

Reviewer: ```suggestion
* [Nullifier inclusion](#nullifier-inclusion)
```

> * [Notes that have not been nullified](#note-validity)

Reviewer: ```suggestion
* [Note validity](#note-validity)
```

> * [Public values](#public-value-inclusion)

Reviewer: ```suggestion
* [Existence of public value](#public-value-inclusion)
```

> * [Contracts](#contract-inclusion)

Reviewer: ```suggestion
* [Contract inclusion](#contract-inclusion)
```

> Using this library, you can check that specific notes or nullifiers happened at specific blocks. This can be useful for things such as:

Reviewer: ```suggestion
Using this library, you can check that specific notes or nullifiers were part of Aztec network state at specific blocks. This can be useful for things such as:
```

> * Verifying a timestamp that was created in a private context

Reviewer: How would this work?

> * Checking eligibility based on historical events (eg for an airdrop) 

Reviewer: ```suggestion
* Checking eligibility based on historical events (e.g. for an airdrop by proving that you owned a note) 
```

* Verifying historic ownership / relinquishing of assets
* Proving existence of a value in public data tree at a given contract slot
* Proving that a contract was deployed in a given block with some parameters

**In this guide you will learn how to**
* Prove a note was included in a specified block

> * Create a nullifier to prove it was not included in a specified block

Reviewer: ```suggestion
* Create a nullifier and prove it was not included in a specified block
```
